synapse/handlers/e2e_keys.py

```diff
@@ -984,56 +980,47 @@ def _get_e2e_cross_signing_verify_key(self, user_id, key_type, from_user_id=None
+        ):
             try:
-                remote_result = yield self.federation.query_client_keys(
-                    user.domain, {"device_keys": {user_id: []}}, timeout=10 * 1000
+                remote_result = yield self.federation.query_user_devices(
+                    user.domain, user_id
                 )
 
-                # Process the result
-                for remote_key_type, remote_user_dict in remote_result.items():
-                    # The key_type variable passed to this function is in the form
-                    # "self_signing","master" etc. whereas the results returned from
-                    # the remote server use "self_signing_keys", "master_keys" etc.
-                    # Remove the "_keys" from the key type
-                    if remote_key_type.endswith("_keys"):
-                        remote_key_type = remote_key_type[:-5]
-
-                    # remote_user_dict is a dictionary in the form of
-                    # {
-                    #   "user_id": {
-                    #     "master_keys": ...
-                    #   },
-                    #   ...
-                    # }
-
-                    # Only extract the keys that pertain to the requested user
-                    key_content_list = remote_user_dict.get(user_id, {}).values()
-
-                    for key_content in key_content_list:
-                        # If the key_type here matches the key we're requesting,
-                        # then this is the key we want to return
-                        if remote_key_type == key_type:
-                            key = key_content
-
-                        # At the same time, save the key to the database for subsequent
-                        # queries
-                        yield self.store.set_e2e_cross_signing_key(
-                            user_id, remote_key_type, key_content
-                        )
-            except (
-                HttpResponseException,
-                NotRetryingDestination,
-                FederationDeniedError,
-                RequestSendFailed,
-            ) as e:
+                # Process each of the retrieved cross-signing keys
+                for key_type in ["master", "self_signing"]:
+                    key_content = remote_result.get(key_type + "_key")
+                    if not key_content:
+                        continue
+
+                    # If this is the desired key type, return it
+                    if key_type == desired_key_type:
+                        key = key_content
+
+                    # At the same time, store this key in the db for
+                    # subsequent queries
+                    yield self.store.set_e2e_cross_signing_key(
+                        user_id, key_type, key_content
+                    )
+            except Exception as e:
                 logger.warning(
                     "Unable to query %s for cross-signing keys of user %s: %s",
                     user.domain,
```

Comment:
I'd recommend logging the type of the exception too, since some exceptions' stringifications make no sense on their own (notably KeyError).

something like:

```
logger.warning("blah blah: %s %s", type(e), e)
```